src/app/interaction-model/MetadataTypes.h

```diff
@@ -0,0 +1,156 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <cstdint>
+#include <optional>
+
+#include <access/Privilege.h>
+#include <app/ConcreteAttributePath.h>
+#include <app/ConcreteClusterPath.h>
+#include <app/ConcreteCommandPath.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/BitFlags.h>
+
+namespace chip {
+namespace app {
+namespace InteractionModel {
+
+enum class ClusterQualityFlags : uint32_t
+{
+    kDiagnosticsData = 0x0001, // `K` quality, may be filtered out in subscriptions
+};
+
+struct ClusterInfo
+{
+    DataVersion dataVersion; // current cluster data version,
+    BitFlags<ClusterQualityFlags> flags;
+
+    /// Constructor that marks data version as mandatory
+    /// for this structure.
+    ClusterInfo(DataVersion version) : dataVersion(version) {}
+};
+
+struct ClusterEntry
+{
+    ConcreteClusterPath path;
+    ClusterInfo info;
+
+    bool IsValid() const { return path.HasValidIds(); }
+
+    static const ClusterEntry kInvalid;
+};
+
+enum class AttributeQualityFlags : uint32_t
+{
+    kListAttribute   = 0x0004, // This attribute is a list attribute
+    kFabricScoped    = 0x0008, // 'F' quality on attributes
+    kFabricSensitive = 0x0010, // 'S' quality on attributes
+    kChangesOmitted  = 0x0020, // `C` quality on attributes
+    kTimed           = 0x0040, // `T` quality on attributes (writes require timed interactions)
+};
+
+struct AttributeInfo
+{
+    BitFlags<AttributeQualityFlags> flags;
+
+    // read/write access will be missing if read/write is NOT allowed
+    std::optional<Access::Privilege> readPrivilege;  // generally defaults to View if readable
+    std::optional<Access::Privilege> writePrivilege; // generally defaults to Operate if writable
+};
+
+struct AttributeEntry
+{
+    ConcreteAttributePath path;
+    AttributeInfo info;
+
+    bool IsValid() const { return path.HasValidIds(); }
+
+    static const AttributeEntry kInvalid;
+};
+
+enum class CommandQualityFlags : uint32_t
+{
+    kFabricScoped    = 0x0001,
+    kFabricSensitive = 0x0002,
+    kTimed           = 0x0004, // `T` quality on commands
```

Comment:
Just as a warning, there are some new ones that are going to show up for TCP as well. Not that this is a blocker, but just an FYI that this list will probably get larger.